cdm/src/main/java/thredds/crawlabledataset/s3utils/S3Helper.java

```diff
@@ -0,0 +1,121 @@
+package thredds.crawlabledataset.s3utils;
+
+import net.sf.ehcache.*;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.*;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+
+public class S3Helper {
+    private static String S3_PREFIX = "s3://";
+    private static String S3_DELIMITER = "/";
+    private static HashMap<String, File> fileStore = new HashMap<String, File>();
+    private static org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(S3Helper.class);
+
+    public String concat(String parent, String child)
+    {
+        if (child.isEmpty())
+            return parent;
+        else
+            return parent + S3_DELIMITER + removeTrailingSlash(child);
+    }
+
+    public String parent(String uri)
+    {
+        int delim = uri.lastIndexOf(S3_DELIMITER);
+        return uri.substring(0, delim);
+    }
+
+    public String delimiter() { return S3_DELIMITER; }
+
+    public String basename(String uri)
+    {
+        return new File(uri).getName();
+    }
+
+    public String stripPrefix(String key, String prefix)
+    {
+        return key.replaceFirst(prefix, "");
+    }
+
+    public String removeTrailingSlash(String str)
+    {
+        if (str.endsWith(S3_DELIMITER))
+            str = str.substring(0, str.length() - 1);
+
+        return str;
+    }
+
+    public File createTempFile(String uri) throws IOException
+    {
+        // We have to save the key twice, as Ehcache will not provide
+        // us with tmpFile when eviction happens, so we use fileStore
+        Path tmpDir = Files.createTempDirectory("S3Download_");
+        String fileBasename = new File(uri).getName();
+        File file = new File(tmpDir.toFile(), fileBasename);
+        file.deleteOnExit();
+        fileStore.put(uri, file);
+        return file;
+    }
+
+    public void deleteFileElement(Element element)
+    {
+        File file = (File) fileStore.get(element.getObjectKey());
+        if (null == file)
+            return;
+
+        // Should cleanup what createTempFile has created, meaning we have
+        // to get rid of both the file and its containing directory
+        file.delete();
+        file.getParentFile().delete();
+
+        fileStore.remove(element.getObjectKey());
+    }
+
+    public File getS3File(String uri, Ehcache cache)
```

Comment:
Perhaps split this class into two - a static helper class and an s3 cache manager